probe-app: initial implementation and description

abonislawski opened the pull request:
This app will be used to parse probes data and extract wave files

Probes will extract data for several probe points in one stream with extra headers. This app will read the resulting file, strip the headers and create wave files for each extracted buffer.

Usage to parse data and create wave files: ./sof-probes -p data.bin

Signed-off-by: Adrian Bonislawski <adrian.bonislawski@linux.intel.com>

dbaluta commented:
@akloniex please put the PR description also in the commit message! After the PR is merged people will be mostly looking inside the commit message logs. They might not even use github interface.

lgirdwood commented on tools/probes/probes_main.c:
```diff
@@ -0,0 +1,96 @@
+
+enum connection_purpose {
+	EXTRACT = 0,
+	INJECT,
```
Can we prefix these.

abonislawski replied:
Sure

lgirdwood commented on tools/probes/probes_main.c:
```diff
@@ -0,0 +1,96 @@
+	exit(0);
+}
+
+void connect_probe(uint32_t buffer_id, enum connection_purpose purpose)
```
I would also expect these to pass some private data structure with context.

abonislawski replied:
These connect and disconnect functions are very small and dont need any structures. I will pass some structures to main parse function when some config options will be added later

lgirdwood commented on tools/probes/probes_main.c:
```diff
@@ -0,0 +1,96 @@
+// Copyright(c) 2018 Intel Corporation. All rights reserved.
+//
+// Author: Adrian Bonislawski <adrian.bonislawski@intel.com>
+
```
Some description in comments on what this tool does.

abonislawski replied:
Done

lgirdwood commented on tools/probes/probes_main.c:
```diff
@@ -0,0 +1,96 @@
+{
+}
+
+void disconnect_probe(uint32_t buffer_id)
```
and return any errors.

abonislawski replied:
done

mmaka1 commented on tools/probes/wave.h:
```diff
@@ -0,0 +1,45 @@
+ * Author: Adrian Bonislawski <adrian.bonislawski@intel.com>
+ */
+
+#ifndef __WAVE_H__
```
I hope we do not need to redefine all this wav format related stuff here. Are there any existing headers in packages we already depend on available?

jajanusz replied:
@mmaka1 I've seen that every audio processing lib  just redefine it over and over and I'm not aware of any 'standard' simple wav lib + it needs suitable license. If you know any lib feel free to share.

lgirdwood replied:
Fwiw, ALSA (aplay) also has a definition of this too, seems like it's redefined everywhere.

abonislawski commented on tools/probes/probes_main.c:
```diff
@@ -0,0 +1,171 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2018 Intel Corporation. All rights reserved.
```
fixed

abonislawski commented:
> I thought the probes were using the compressed API to inject/retrieve data from ring buffers.
> What's the deal with wave files?

Compressed stream (from probes in FW) will contain several probe points so this app will read this one resulting file, strip the probe data_packets headers, generate wave headers for new files and copy data to corresponding wave files

abonislawski commented:
Initial probe code was merged to FW so I have updated app code, it will now include FW headers, crc32 is also used.

I have also deleted some optional features to make it simpler to review and merge, more optional features will be added in separate PRs

jajanusz commented on tools/probes/probes_main.c:
```diff
@@ -0,0 +1,288 @@
+	CHECK
+};
+
+uint32_t sample_rate[15] = {
```
Why no `static`? It is used only in this file. Also I guess `15` is unnecessary, compiler should fill it.

abonislawski replied:
Fixed

jajanusz commented on tools/probes/probes_main.c:
```diff
@@ -0,0 +1,288 @@
+{
+	int opt;
+
+	while ((opt = getopt(argc, argv, "d:e:hi:p:")) != -1) {
```
It looks that most of these args are unused (just p & h are in use atm).

abonislawski replied:
sorry, forgot about this

jajanusz commented on tools/probes/wave.h:
```diff
@@ -0,0 +1,45 @@
+#define HEADER_RIFF 0x46464952
+#define HEADER_WAVE 0x45564157
+#define HEADER_FMT  0x20746d66
+#define HEADER_DATA 0x61746164
```
I know that these are just 4-chars array converted to ints, but it may not be obvious for everyone, so you can add some note.

abonislawski replied:
added comments with ascii explanatory

lgirdwood commented on tools/probes/probes_main.c:
```diff
@@ -0,0 +1,288 @@
+#define APP_NAME "sof-probes"
+
+#define DATA_READ_LIMIT 1000
+#define FILES_LIMIT	8
```
can we describe these macros with some comments.

abonislawski replied:
done

lgirdwood commented on tools/probes/probes_main.c:
```diff
@@ -0,0 +1,288 @@
+	READY = 0,
+	SYNC,
+	CHECK
+};
```
some comments here describing each state would be useful.

abonislawski replied:
added

lgirdwood commented on tools/probes/probes_main.c:
```diff
@@ -0,0 +1,288 @@
+{
+	uint32_t i, chunk_size;
+
+	for (i = 0; i < FILES_LIMIT; i++) {
```
what are we doing here ?

abonislawski replied:
please check if these comments are ok

lgirdwood commented on tools/probes/probes_main.c:
```diff
@@ -0,0 +1,288 @@
+			files[i].size *= 8;
+			chunk_size = files[i].size + 36;
+
+			fseek(files[i].fd, sizeof(uint32_t), SEEK_SET);
```
and here, lots of magic numbers too

abonislawski replied:
Found a way to get rid of them but not sure if it is more clearer now, please check

lgirdwood commented on tools/probes/probes_main.c:
```diff
@@ -0,0 +1,288 @@
+	if (received_crc == calc_crc)
+		return 1;
+	else
+		return 0;
```
In general we should be returning 0 on success and negative error numbers. So we should return -EINVAL here and we could also write something to stderr informing the user.

abonislawski replied:
Right, fixed

lgirdwood commented on tools/probes/probes_main.c:
```diff
@@ -0,0 +1,288 @@
+	uint32_t total_data_to_copy = 0;
+	uint32_t data_to_copy = 0;
+	uint32_t *w_ptr;
+
```
newline not needed.

abonislawski replied:
deleted

lgirdwood commented on tools/probes/probes_main.c:
```diff
@@ -0,0 +1,288 @@
+
+	int i, j, file;
+
+	enum p_state state = 0;
```
please use the ENUM here when satting state rather than an int.

abonislawski replied:
done

lgirdwood commented on tools/probes/probes_main.c:
```diff
@@ -0,0 +1,288 @@
+	memset(&files, 0, sizeof(struct wave_files) * FILES_LIMIT);
+	file = -1;
+
+	do {
```
what does this do ?

abonislawski replied:
added bunch of comments in this loop, please check if they are ok

lgirdwood commented on tools/probes/probes_main.c:
```diff
@@ -0,0 +1,288 @@
+					free(packet);
+					exit(0);
+				}
+				packet = malloc(sizeof(struct probe_data_packet));
```
if we are only processing one packet at a time then we should allocate it once at init.

lgirdwood replied:
This part needs commenting too.

abonislawski replied:
Not possible because we dont know how big will be this packet. We can for example at this stage have loaded only half of probe_data_packet so the size is unknown

lgirdwood replied:
just allocate MAX packet size otherwise we have an extra kernel context switch for every packet

abonislawski replied:
Sorry, I misunderstood your first comment. One allocation at init is of course good idea, added with PACKET_MAX_SIZE

plbossart commented on tools/probes/probes_main.c:
```diff
@@ -0,0 +1,306 @@
+
+#define PACKET_MAX_SIZE	4096	/**< Size limit for probe data packet */
+#define DATA_READ_LIMIT 1024	/**< Data limit for file read */
+#define FILES_LIMIT	8	/**< Maximum num of files */
```
shouldn't this be aligned with the max number of extractors?

abonislawski replied:
yes, it should but it is generated by kconfig in build dir which is unknown for this app. Maybe I will just increase this limit, ok?

plbossart commented on tools/probes/CMakeLists.txt:
```diff
@@ -0,0 +1,18 @@
+# SPDX-License-Identifier: BSD-3-Clause
```
comment on the commit message: can you clarify the usage? It seems to me this is an off-line tool to split the probe data into wave files, but how do we get the probe data in the first place?

abonislawski replied:
Yes, offline tool to split the probe data into wave files. This PR is only about this tool, how do we get the probe data depends on kernel implementation (user will use debugfs to add probe points etc.) which probably will be explained in docs.

For this tool the usage is very simple: 
"Probes will extract data for several probe points in one stream
 with extra headers. This app will read the resulting file,
 strip the headers and create wave files for each extracted buffer.

 Usage to parse data and create wave files: ./sof-probes -p data.bin"

What more would you like to add here?

lgirdwood commented on tools/probes/probes_main.c:
```diff
@@ -0,0 +1,306 @@
+
+#define PACKET_MAX_SIZE	4096	/**< Size limit for probe data packet */
+#define DATA_READ_LIMIT 1024	/**< Data limit for file read */
+#define FILES_LIMIT	32	/**< Maximum num of files */
```
what files ? probe output files, injector input files ? or both ?

abonislawski replied:
probe output files, fixed

lgirdwood commented on tools/probes/probes_main.c:
```diff
@@ -0,0 +1,306 @@
+
+int init_wave(struct wave_files *files, uint32_t buffer_id, uint32_t format)
+{
+	char path[64];
```
This should be a macro and probably bigger

abonislawski replied:
done

abonislawski commented:
> One more thing, what happens to any unwritten file data if I press Ctrl-C ? Do we need a signal handler to call finalise() and flush all files as I suspect this will be how we exit the tool (as it loops forever of data).

@lgirdwood not a problem because it doesnt loops forever :)  
Usage to parse data and create wave files: ./sof-probes -p data.bin
At this stage data.bin is already closed file with all extraction data, so you just run sof-probes and its like 1 second operation

lgirdwood commented:
> > One more thing, what happens to any unwritten file data if I press Ctrl-C ? Do we need a signal handler to call finalise() and flush all files as I suspect this will be how we exit the tool (as it loops forever of data).
> 
> @lgirdwood not a problem because it doesnt loops forever :)
> Usage to parse data and create wave files: ./sof-probes -p data.bin
> At this stage data.bin is already closed file with all extraction data from FW/kernel, so you just run sof-probes and its like 1 second operation

ok, but I would expect the reader tool to read(forever) until stopped by the user. That way we can also probe pause/release and start/stop conditions ?

abonislawski commented:
> ok, but I would expect the reader tool to read(forever) until stopped by the user. That way we can also probe pause/release and start/stop conditions ?

You can probe pause release, start/stop. :)
Because you can play with your streams as always, this app is the last step for probe extraction flow,  its not intended for real-time work

lgirdwood commented:
@abonislawski can you check CI

abonislawski commented:
@lgirdwood I think this app has no effect on FW and CI
updated braces for checkpack check
